be/src/storage/binlog_file_writer.cpp

```diff
@@ -0,0 +1,457 @@
+}
+
+Status BinlogFileWriter::prepare(int64_t version, const RowsetId& rowset_id, int64_t start_seq_id,
+                                 int64_t change_event_timestamp_in_us) {
+    CHECK(_writer_state == WAIT_WRITE);
+
+    PendingVersionContext* version_context = _pending_version_context.get();
+    version_context->version = version;
+    version_context->rowset_id.init(rowset_id.to_string());
+    version_context->start_seq_id = start_seq_id;
+    version_context->change_event_timestamp_in_us = change_event_timestamp_in_us;
+    version_context->num_pages = 0;
+    version_context->rowsets.clear();
+    version_context->rowsets.emplace(rowset_id);
+
+    PendingPageContext* page_context = _pending_page_context.get();
+    page_context->start_seq_id = start_seq_id;
+    // start_seq_id > end_seq_id means there is no log entry
+    page_context->end_seq_id = start_seq_id - 1;
+    page_context->num_log_entries = 0;
+    page_context->last_segment_index = -1;
+    page_context->last_row_id = -1;
+    page_context->estimated_page_size = 0;
+    page_context->page_header.Clear();
+    page_context->page_content.Clear();
+    page_context->rowsets.emplace(rowset_id);
+
+    _writer_state = PREPARE;
+    return Status::OK();
+}
+
+Status BinlogFileWriter::add_empty() {
+    CHECK(_writer_state == PREPARE);
+    CHECK(_pending_page_context->num_log_entries == 0)
+            << "Empty rowset should only have one empty log entry,"
+            << ", version " << _pending_version_context->version << ", file id " << _file_id << ", file name "
+            << _file_path << ", actual number of entries " << _pending_page_context->num_log_entries;
+
+    LogEntryPB* log_entry = _pending_page_context->page_content.add_entries();
+    log_entry->set_entry_type(EMPTY_PB);
+
+    // No need to add rowset
+    _pending_version_context->rowsets.clear();
+    _pending_page_context->rowsets.clear();
+    PendingPageContext* page_context = _pending_page_context.get();
+    page_context->num_log_entries += 1;
+    // TODO reduce estimation cost
+    page_context->estimated_page_size += log_entry->ByteSizeLong();
+    return Status::OK();
+}
+
+Status BinlogFileWriter::add_insert_range(int32_t seg_index, int32_t start_row_id, int32_t num_rows) {
+    CHECK(_writer_state == PREPARE);
+    RETURN_IF_ERROR(_switch_page_if_full());
+    PendingPageContext* page_context = _pending_page_context.get();
+    LogEntryPB* log_entry = page_context->page_content.add_entries();
+    log_entry->set_entry_type(INSERT_RANGE_PB);
+    InsertRangePB* entry_data = log_entry->mutable_insert_range_data();
+    bool in_one_segment = false;
+    // if the last and current log entries are in the same segment, no need to set file id.
+    // When reading and iterating the page, we can get the file id from the last log entry
+    if (page_context->last_segment_index != seg_index) {
+        _set_file_id_pb(_pending_version_context->rowset_id, seg_index, entry_data->mutable_file_id());
```

Comment:
how to make sure the original data in rowset_id and seg_index still exist after compaction?
where to realize this logic?